File: python/xss/template_injection.py
from flask import Flask, request, render_template_string
from jinja2 import Template

app = Flask(__name__)


# VULN 1: Server-Side Template Injection - user input rendered as Jinja2 template
@app.route("/email/preview")
def preview_email_template():
    template_str = request.args.get("template", "Hello {{ name }}")
    rendered = render_template_string(template_str)
    return rendered


# VULN 2: SSTI via product description field rendered as template
@app.route("/product/<int:product_id>/description")
def render_product_description(product_id):
    description = get_product_description(product_id)  # user-supplied content
    return render_template_string(description)


# VULN 3: SSTI in promotional banner - admin user controls template string
@app.route("/banner/render", methods=["POST"])
def render_banner():
    banner_template = request.json.get("template")
    context = {"discount": 20, "code": "SAVE20"}
    t = Template(banner_template)
    return t.render(**context)


def get_product_description(product_id):
    return "Great product!"

File: ruby/xss/template_injection.rb
require 'sinatra'
require 'erb'
require 'erubi'
require 'liquid'

# VULN 1: ERB template injection - user controls template string, gains full Ruby code execution
post '/products/custom_description' do
  user_template = params[:template]
  product = Product.find(params[:product_id])
  # ERB.new(user_template).result(binding) executes arbitrary Ruby if template is user-controlled
  rendered = ERB.new(user_template).result(binding)
  { description: rendered }.to_json
end

# VULN 2: Erubi rendering user input as template - arbitrary Ruby execution via <%= %> tags
post '/emails/preview' do
  email_template = params[:email_body]
  order = Order.find(params[:order_id])
  # Erubi::Engine compiles and evaluates the template, user can inject <%= `id` %>
  src = Erubi::Engine.new(email_template).src
  rendered = eval(src)
  { preview: rendered }.to_json
end

# VULN 3: Liquid template with user-controlled template string - object traversal and data exfiltration
post '/invoices/generate' do
  invoice_template = params[:template]
  order = Order.find(params[:order_id])
  # User-controlled Liquid template allows traversing exposed objects and leaking data
  template = Liquid::Template.parse(invoice_template)
  rendered = template.render('order' => order, 'user' => order.user, 'config' => Rails.application.config)
  { invoice: rendered }.to_json
end
